src/lang-promql/client/prometheus.ts

```diff
@@ -279,3 +203,124 @@ export class HTTPPrometheusClient implements PrometheusClient {
       });
   }
 }
+
+class Cache {
+  // completeAssociation is the association between a metric name, a label name and the possible label values
+  private readonly completeAssociation: Map<string, Map<string, Set<string>>>;
+  // metricMetadata is the association between a metric name and the associated metadata
+  private readonly metricMetadata: Map<string, MetricMetadata[]>;
+  private labelValues: Map<string, string[]>;
+  private labelNames: string[];
+
+  constructor() {
+    this.completeAssociation = new Map<string, Map<string, Set<string>>>();
+    this.metricMetadata = new Map<string, MetricMetadata[]>();
+    this.labelValues = new Map<string, string[]>();
+    this.labelNames = [];
+  }
+
+  setAssociations(metricName: string, series: Map<string, string>[]): void {
+    buildAssociations(metricName, series, this.completeAssociation);
+  }
+
+  setMetricMetadata(metricName: string, metadata: MetricMetadata[]): void {
+    this.metricMetadata.set(metricName, metadata);
+  }
+
+  getMetricMetadata(): Map<string, MetricMetadata[]> {
+    return this.metricMetadata;
+  }
+
+  setLabelNames(labelNames: string[]): void {
+    this.labelNames = labelNames;
+  }
+
+  getLabelNames(metricName?: string): string[] {
+    if (!metricName || metricName.length === 0) {
+      return this.labelNames;
+    }
+    const labelSet = this.completeAssociation.get(metricName);
+    return labelSet ? Array.from(labelSet.keys()) : [];
+  }
+
+  setLabelValues(labelName: string, labelValues: string[]): void {
+    this.labelValues.set(labelName, labelValues);
+  }
+
+  getLabelValues(labelName: string, metricName?: string): string[] {
+    if (!metricName || metricName.length === 0) {
+      const result = this.labelValues.get(labelName);
+      return result ? result : [];
+    }
+
+    const labelSet = this.completeAssociation.get(metricName);
+    if (labelSet) {
+      const labelValues = labelSet.get(labelName);
+      return labelValues ? Array.from(labelValues) : [];
+    }
+    return [];
+  }
+}
+
+export class CachedPrometheusClient implements PrometheusClient {
+  private readonly cache: Cache;
+  private readonly client: PrometheusClient;
+
+  constructor(client: PrometheusClient) {
+    this.client = client;
+    this.cache = new Cache();
+  }
+
+  labelNames(metricName?: string): Promise<string[]> {
+    if (this.cache.getLabelNames(metricName) && this.cache.getLabelNames(metricName).length > 0) {
```

Comment:
right now, yeah this condition is useless.

But I'm wondering if we could optimize a bit what we are storing specially concerning the metric.
The metrics are currently stored in the attribute `labelValues` with the key `__name__`. Don't know if it's worth to store them directly in the attribute `completeAssociation`. Maybe it will cost too much in term of empty struct created (i.e. for every metricName, you would have an empty Map associated.)

Just by writing it, I feel less confident hahaha. But in my head it didn't look so bad ...

Reply:
actually the 2nd condition is happening when the cache is empty. If you don't check that the array is empty, then it never makes the HTTP request

Reply:
Ahh, yes. Makes sense, ok.